# PREDICTION DE LA MALADIE CARDIAQUE ETAPE DU PREPROCESSING ET DE LA SELECTION D'UN MODELE DE MACHINE LEARNING

Ceci est la suite de la notre démarche de modélisation, dans laquelle au niveau de la première partie, nous avons effectué une analyse exploratoire sur les variables, aujourd'hui nous poursuivons en effectuant un preprocessing sur les variables en vue de choisir un meilleur modèle de machine learning 

## A) Rappel des résultats obtenus lors de l'analyse exploratoire

Ici nous rappelons quelques principaux résultats obténus lors de l'analyse exploratoire
                - jeu de donnée de 918 individus et 12 variables avec aucune donnée manquante
                - population adulte (40-60 ans), parmi eux bon nombre sont des hommes ont un électrocardiogramme normal, une
                  douleur thoracique asymtomatique, une pente de segment plat, un taux de cholestérol élévé, une présence de
                  maladie cardique environ 60% d'entre eux, une absence d'angime, une pression arterielle élevé
                - cinq variables de type objet soit le sexe, le type de couleur thoracique, le résultats de             
                  l'électrocardiogramme au repos, l'angine, la pente du segment d'exercice et le reste des variables de type 
                  numérique
                  
A ce niveau l'on essayera avant de s'y lancer dans le preprocessing de donner la nature de chaque variable. Mais avant tous, dans la ligne qui suit nous importerons les differentes librairies que nous utiliserons dans cet exercice.
                  

                
               

In [71]:
import pandas as pd #librairie pandas me permettant de faire des manip sur des tableaux
import matplotlib.pyplot as plt #librairie me permettant de faire des visualisations graphiques
import seaborn as sns #alternative à la librairie précedente
from sklearn.preprocessing import LabelEncoder # cette librairie nous permettra d'encoder les variables qualitatives lors du préprocessing
from sklearn.preprocessing import OrdinalEncoder # cette librairie nous permettra d'encoder les variables qualitatives lors du préprocessing
from sklearn.preprocessing import OneHotEncoder # cette librairie nous permettra d'encoder les variables qualitatives lors du préprocessing

In [26]:
df = pd.read_excel("heart_disease.xlsx") #comme nous l'avons dis, nous avons affaire à un fichier excel nous utilisons donc 
                                         # la librairie pandas et l'attribut read_excel puis le nom du fichier
    
df.head(10) #pour verifier si le jeu de donnée a bien été chargé 

AGE   SEXE  TDT  PAR  CHOLESTEROL  GAJ     ECG  FCMAX ANGINE  DEPRESSION   \
0   40  homme   AA  140          289    0  Normal    172    Non          0.0   
1   49  femme  DNA  160          180    0  Normal    156    Non          1.0   
2   37  homme   AA  130          283    0      ST     98    Non          0.0   
3   48  femme  ASY  138          214    0  Normal    108    Oui          1.5   
4   54  homme  DNA  150          195    0  Normal    122    Non          0.0   
5   39  homme  DNA  120          339    0  Normal    170    Non          0.0   
6   45  femme   AA  130          237    0  Normal    170    Non          0.0   
7   54  homme   AA  110          208    0  Normal    142    Non          0.0   
8   37  homme  ASY  140          207    0  Normal    130    Oui          1.5   
9   48  femme   AA  120          284    0  Normal    120    Non          0.0   

       PENTE  CŒUR  
0  Ascendant     0  
1       Plat     1  
2  Ascendant     0  
3       Plat     1  
4  Ascendant     0  
5  Ascendant     0  
6  Ascendant     0  
7  Ascendant     0  
8       Plat     1  
9  Ascendant     0

#### B) Nature des variables 
B-1)Nature des variables 
rappelons que nous avons dans le dataset cinq variables de type objet et sept variables de type numérique. Regardons de près les valeurs que chaque variable peut prendre

In [27]:
vquali = df.select_dtypes(object) # pour récupérer les variables qualitatives dans un tableau
vquanti = df.select_dtypes(exclude=['object']) # pour récuperer les variables quantitatives 
for col in vquali.select_dtypes(object):   # fonction pour voir la valeur que peux prendre chaque variable qualitative
    print(vquali[col].value_counts())

homme    725
femme    193
Name: SEXE, dtype: int64
ASY    496
DNA    203
AA     173
AT      46
Name: TDT, dtype: int64
Normal    552
LVH       188
ST        178
Name: ECG, dtype: int64
Non    547
Oui    371
Name: ANGINE, dtype: int64
Plat          460
Ascendant     395
Descendant     63
Name: PENTE, dtype: int64


### Affichage du Descriptif du jeu de donnée
AGE : âge du patient 
SEXE : sexe du patient [homme, femme]
TDT : type de douleur thoracique [AT : angine typique, AA : angine atypique, DNA : douleur non angineuse, ASY : asymptomatique]
PAR : pression artérielle au repos [mm Hg]
CHOLESTEROL : cholestérol sérique [mm/dl]
GAJ : glycémie à jeun [1 : si glycémie à jeun > 120 mg/dl, 0 : sinon]
ECG au repos : résultats de l'électrocardiogramme au repos [Normal : normal, ST : présentant une anomalie de l'onde ST-T , HVG : montrant une hypertrophie ventriculaire gauche probable]
FCMAX : fréquence cardiaque maximale atteinte 
ANGINE : angine induite par l'exercice [Oui, Non]
DEPRESSION : Valeur numérique mesurée en dépression
PENTE: la pente du segment ST d'exercice de pointe [Ascendant, Plat, Descendant]
COEUR : maladie cardiaque[1 : Présence de la maladie, 0 : Absence de la maladie]

### Conclusion partielle
Après observation nous remarquons que nous avons des variables qualitatives nominales.
Ces variables qualitatives nominales seront codées par du one hot encoding en vue de en vue de n'attribuer aucun poids

### C) Encodage des variables qualitatives
Rappelons que nous avions separez les données qualitatives dans vquali

In [28]:

encoder = OneHotEncoder(sparse=False) # l'option sparse pour afficher dans une liste
enco = encoder.fit_transform(vquali)
encode = pd.DataFrame(enco, index = vquali.index)
encode.head()

0    1    2    3    4    5    6    7    8    9    10   11   12   13
0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  1.0  0.0  0.0
1  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0
2  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0
3  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0
4  0.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  1.0  0.0  1.0  0.0  0.0

In [29]:
encode.columns = encoder.get_feature_names(vquali.columns)
encode.head()

SEXE_femme  SEXE_homme  TDT_AA  TDT_ASY  TDT_AT  TDT_DNA  ECG_LVH  \
0         0.0         1.0     1.0      0.0     0.0      0.0      0.0   
1         1.0         0.0     0.0      0.0     0.0      1.0      0.0   
2         0.0         1.0     1.0      0.0     0.0      0.0      0.0   
3         1.0         0.0     0.0      1.0     0.0      0.0      0.0   
4         0.0         1.0     0.0      0.0     0.0      1.0      0.0   

   ECG_Normal  ECG_ST  ANGINE_Non  ANGINE_Oui  PENTE_Ascendant  \
0         1.0     0.0         1.0         0.0              1.0   
1         1.0     0.0         1.0         0.0              0.0   
2         0.0     1.0         1.0         0.0              1.0   
3         1.0     0.0         0.0         1.0              0.0   
4         1.0     0.0         1.0         0.0              1.0   

   PENTE_Descendant  PENTE_Plat  
0               0.0         0.0  
1               0.0         1.0  
2               0.0         0.0  
3               0.0         1.0  
4               0.0         0.0

In [49]:
df1 = pd.concat([encode, vquanti], axis = 1)
df1

SEXE_femme  SEXE_homme  TDT_AA  TDT_ASY  TDT_AT  TDT_DNA  ECG_LVH  \
0           0.0         1.0     1.0      0.0     0.0      0.0      0.0   
1           1.0         0.0     0.0      0.0     0.0      1.0      0.0   
2           0.0         1.0     1.0      0.0     0.0      0.0      0.0   
3           1.0         0.0     0.0      1.0     0.0      0.0      0.0   
4           0.0         1.0     0.0      0.0     0.0      1.0      0.0   
..          ...         ...     ...      ...     ...      ...      ...   
913         0.0         1.0     0.0      0.0     1.0      0.0      0.0   
914         0.0         1.0     0.0      1.0     0.0      0.0      0.0   
915         0.0         1.0     0.0      1.0     0.0      0.0      0.0   
916         1.0         0.0     1.0      0.0     0.0      0.0      1.0   
917         0.0         1.0     0.0      0.0     0.0      1.0      0.0   

     ECG_Normal  ECG_ST  ANGINE_Non  ...  PENTE_Ascendant  PENTE_Descendant  \
0           1.0     0.0         1.0  ...              1.0               0.0   
1           1.0     0.0         1.0  ...              0.0               0.0   
2           0.0     1.0         1.0  ...              1.0               0.0   
3           1.0     0.0         0.0  ...              0.0               0.0   
4           1.0     0.0         1.0  ...              1.0               0.0   
..          ...     ...         ...  ...              ...               ...   
913         1.0     0.0         1.0  ...              0.0               0.0   
914         1.0     0.0         1.0  ...              0.0               0.0   
915         1.0     0.0         0.0  ...              0.0               0.0   
916         0.0     0.0         1.0  ...              0.0               0.0   
917         1.0     0.0         1.0  ...              1.0               0.0   

     PENTE_Plat  AGE  PAR  CHOLESTEROL  GAJ  FCMAX  DEPRESSION   CŒUR  
0           0.0   40  140          289    0    172          0.0     0  
1           1.0   49  160          180    0    156          1.0     1  
2           0.0   37  130          283    0     98          0.0     0  
3           1.0   48  138          214    0    108          1.5     1  
4           0.0   54  150          195    0    122          0.0     0  
..          ...  ...  ...          ...  ...    ...          ...   ...  
913         1.0   45  110          264    0    132          1.2     1  
914         1.0   68  144          193    1    141          3.4     1  
915         1.0   57  130          131    0    115          1.2     1  
916         1.0   57  130          236    0    174          0.0     1  
917         0.0   38  138          175    0    173          0.0     0  

[918 rows x 21 columns]

## C) mise en place du modèle de machine learning et machine learning avancé

nous allons appliquer une régression logistique comparé à un séparateur à vaste marge. Cette partie est réservée pour le troisième notebook. En effet le préprocessing nous permet d'otenir le meilleur modèle à partir de certains paramètres dans le modèle. Dans ce troisième notebook, nous mettrons un modèle de machine learning de base et évaluerons les performances avec d'autres modèles

### C'est ici que prends fin l'exercice